### PR TITLE
3406 ab testing architecture docs

### PR DESCRIPTION
**Pull request by kpethtel**
**Story card:** [ch3406](https://app.clubhouse.io/simpledotorg/story/3406/write-architecture-doc)

## Because

This is an attempt to document the A/B framework. The ticket also calls for diagrams, but I'm going to have to defer that for a future ticket. This is my first architecture doc at Simple, so please feel free to let me know what you'd like to see.

### CODE REVIEW

File: doc/arch/016-ab-testing.md [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+
+Those two types of experiments will require process differences. When the patient has an upcoming appointment, the reminders must be sent relative to the appointment date. Messages can be sent before, on, or after the appointment date.
+
+That doesn't make sense for patients who do not have upcoming appointments. Messages for stale patients will start on the same day the patient is added to the experiment. There are many patients who fall into this category, and we don't want to risk driving too many patients back to care at once and overwhelming clinics. To mitigate that possibility, we will add 10,000 patients per day to the experiment and schedule their reminders to be sent starting on the day they were added to the experiment. Because patients can receive no reminders, a single reminder, or a series of reminders, the actual number of reminders sent on any given day may exceed 10,000.
```
Comment (rsanheim) [inline doc/arch/016-ab-testing.md:19]:
```suggestion
Messaging for stale patients will run on a continual daily basis during the experiment. This is to stagger the messages over the experiment time so as to not overwhelm clinics with many patients coming to a clinic on the same day.  We will add a random batch of stale patients per day (currently 10,000) to the experiment and schedule them based on the messaging templates configured in the experiment.
```

File: doc/arch/016-ab-testing.md [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+
+In both types of experiments, patients will be randomly placed into treatment groups. Treatment groups will define the message texts, number of messages, and when to send the messages. Test patients will not receive the pre-existing text message sent three days after they miss an appointment. Control group patients will receive no messages at all.
```
Comment (rsanheim) [inline doc/arch/016-ab-testing.md:21]:
```suggestion
In both types of experiments, patients will be randomly placed into treatment groups. Treatment groups will define the message texts, number of messages, and when to send the messages. Patients in an experiment will not receive the standard missed visit reminder that is sent three days after a scheduled appointment. Control group patients will receive no messages at all.
```

trying to clarify this point

File: doc/arch/016-ab-testing.md [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+
+## Patient Selection Criteria
+
+Simple servers are hosted by country and the selection pool will incude all patients on the server, so all patients in the experiment country will be potentially eligible.
+
+All patients must meet the following criteria for inclusion:
+- at least 18 years old
+- hypertensive
+- has a phone capable of receiving text messages. We verify this via Exotel.
+- is not in an ongoing experiment or one that ended within the past 14 days.
```
Comment (rsanheim) [inline doc/arch/016-ab-testing.md:31]:
```suggestion
- is not in an ongoing experiment or one that ended recently
```

less about implementation, more about the intention

**Review by rsanheim — CHANGES_REQUESTED** [review]
Hey this is a good start.  Overall I think its a bit too low level for an ADR, and could use some clean up so that it focuses more on the long term implications of moving to an AB testing approach for our notifications.

I left some comments around things to remove and some stuff to expand on, lemme know if you'd like to discuss further!

File: doc/arch/016-ab-testing.md [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+Tamil
+Telugu
+
+## Notification window
+
+We want to ensure that we only send notifications during appropriate hours, so all notifications are scheduled to be sent during the next messaging window (see `Communication.next_messaging_time`). We initially send messages through WhatsApp or Imo, then fall back to SMS. If the callback from WhatsApp/Imo tells us the message failed, we schedule the message to resend either immediately or during the messaging window tomorrow.
+
+Because these experiments will dramatically increase the number of notifications being sent daily, we are expanding the notifications window. Currently, the window is from 10am-4pm IST. Later experiments will likely test sending messages at different times of day.
+
+## Consequences
```
[the remark this answers is not in the document]
Comment (kpethtel) [reply doc/arch/016-ab-testing.md:103]: Valid points, especially re: rate limits. I'll work on it.

**Review by harimohanraj89 — APPROVED** [review]
I like the separation of the ADR and the how-to wiki doc. Let's get this merged.

**Pull request merged**